### ISSUE TITLE
Specialized CNN Cuda Kernels

### ISSUE THREAD

**Issue by haroonsyed** [haroonsyed/Rust-Machine-Learning#40]
While performing the CNN operations using compositions of kernel launches of matrix operations is nice. It is not performant, especially when dealing with packed operations due to the amount of metadata being generated and copied over.

It is creating a mild CPU bottleneck.

To avoid this, I think it is best I compose cuda device functions and have general kernel launches for CNN steps. For example, the cnn_feed_forward cuda kernel launch will have device functions that do correlate_valid, relu etc. 

In summary:
1. Make all basic matrix operations callable via device functions (existing kernel launches can then just call the device functions...they should get inlined with no performance penalty :crossed_fingers: )
2. Call those device function for specialized CNN cuda kernels.
3. Profit from reduced meta data, cpu usage, io usage, kernel launches.

**Comment by haroonsyed** [comment]
#43
